File: cluster-viz/frontend/src/lib/types/graph.ts
// TypeScript mirrors of Go models

export interface GraphNode {
  id: string;
  kind: string;
  name: string;
  namespace: string;
  status: string;
  labels: Record<string, string>;
  parentId?: string;
  metadata: Record<string, string>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  kind: 'ownership' | 'routing' | 'selection' | 'dataflow' | 'parentRef';
  label?: string;
}

export interface Graph {
  nodes: Record<string, GraphNode>;
  edges: Record<string, GraphEdge>;
}

export type EventType = 'SYNC' | 'ADDED' | 'MODIFIED' | 'DELETED';

export interface GraphEvent {
  type: EventType;
  node?: GraphNode;
  edge?: GraphEdge;
  graph?: Graph;
}

File: cluster-viz/backend/pkg/models/graph.go
package models

// GraphNode represents a Kubernetes resource in the graph
type GraphNode struct {
	ID        string            `json:"id"`
	Kind      string            `json:"kind"`
	Name      string            `json:"name"`
	Namespace string            `json:"namespace"`
	Status    string            `json:"status"`
	Labels    map[string]string `json:"labels"`
	ParentID  string            `json:"parentId,omitempty"`
	Metadata  map[string]string `json:"metadata"`
}

// GraphEdge represents a relationship between Kubernetes resources
type GraphEdge struct {
	ID     string `json:"id"`
	Source string `json:"source"`
	Target string `json:"target"`
	Kind   string `json:"kind"` // ownership, routing, selection, dataflow, parentRef
	Label  string `json:"label,omitempty"`
}

// Graph represents the complete cluster topology
type Graph struct {
	Nodes map[string]*GraphNode `json:"nodes"`
	Edges map[string]*GraphEdge `json:"edges"`
}

// NewGraph creates an empty graph
func NewGraph() *Graph {
	return &Graph{
		Nodes: make(map[string]*GraphNode),
		Edges: make(map[string]*GraphEdge),
	}
}
